chain/stmgr/call.go

```diff
@@ -107,22 +107,16 @@ func (sm *StateManager) callInternal(ctx context.Context, msg *types.Message, pr
 		}
 	}
 
-	var vmHeight abi.ChainEpoch
-	if checkGas {
-		// Since we're simulating a future message, pretend we're applying it in the "next" tipset
-		vmHeight = ts.Height() + 1
-		if stateCid == cid.Undef {
-			stateCid, _, err = sm.TipSetState(ctx, ts)
-			if err != nil {
-				return nil, xerrors.Errorf("computing tipset state: %w", err)
-			}
-		}
-	} else {
-		// If we're not checking gas, we don't want to have to execute the tipset like above. This saves a lot of computation time
-		vmHeight = pts.Height() + 1
-		if stateCid == cid.Undef {
-			stateCid = ts.ParentState()
+	// Unless executing on a specific state cid, apply all the messages from the current tipset
+	// first. Unfortunately, we can't just execute the tipset, because that will run cron. We
+	// don't want to apply miner messages after cron runs in a given epoch.
+	if stateCid == cid.Undef {
+		stateCid = ts.ParentState()
+		tsMsgs, err := sm.cs.MessagesForTipset(ctx, ts)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to lookup messages for parent tipset: %w", err)
 		}
+		priorMsgs = append(tsMsgs, priorMsgs...)
```

Comment:
Are we sure this isn't breaking StateCall / `lotus state compute-state`?

Exchanges rely on this to track transactions.

Reply:
That uses https://github.com/filecoin-project/lotus/blob/7aa6663127982aa404e68bdefaadb3f041eb5120/chain/stmgr/utils.go#L68

Reply:
Which is also incorrect? Well, it's correct if you don't pass any messages.

Reply:
Hm. It does affect `StateCall`. I'm going to rework this a bit.

Reply:
Done. I now only do this for `CallWithGas`.